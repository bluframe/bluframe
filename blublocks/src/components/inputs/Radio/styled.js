// @flow
/* eslint-disable no-ternary */

import styled, { type StyledComponent } from "styled-components"
import type { Theme } from "@bluframe/blublocks"

export const Wrapper: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  align-items: center;
  column-gap: 8px;
  cursor: pointer;
  display: grid;
  grid-template-columns: auto 1fr;
`

export const Circle: StyledComponent<
  {| +checked: boolean |},
  Theme,
  HTMLDivElement
> = styled.div`
  border-color: ${({ theme }) => theme.palette.primary.main};
  border-style: solid;
  border-width: ${({ checked }) => (checked ? "2px" : "1px")};
  border-radius: 50%;
  cursor: pointer;
  height: 20px;
  position: relative;
  width: 20px;

  &::after {
    background: ${({ checked, theme }) =>
      checked ? theme.palette.primary.main : "transparent"};
    border-radius: 50%;
    content: "";
    height: 12px;
    left: 2px;
    position: absolute;
    top: 2px;
    width: 12px;
  }
`

export const Input: StyledComponent<
  {||},
  Theme,
  HTMLInputElement
> = styled.input.attrs({ type: "radio" })`
  opacity: 0;
  position: absolute;
`

export const Label: StyledComponent<
  {||},
  Theme,
  HTMLLabelElement
> = styled.label`
  color: ${({ theme }) => theme.palette.text.main};
  font-size: 16px;
`
