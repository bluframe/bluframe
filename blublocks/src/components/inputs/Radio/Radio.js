// @flow
/* eslint-disable no-ternary */

import { Circle, Input, Label, Wrapper } from "./styled"
import type { ComponentProps } from "."
import React from "react"

const Radio = ({
  inputId,
  isChecked,
  label,
  labelId,
  name
}: ComponentProps): React$Node => (
  <Wrapper>
    <Circle
      aria-checked={isChecked}
      aria-labelledby={labelId}
      checked={isChecked}
      id={inputId}
      role="radio"
    />

    <Input defaultChecked={isChecked} name={name} />

    <Label id={inputId ? `${inputId}-label` : ""}>{label}</Label>
  </Wrapper>
)

export default Radio
