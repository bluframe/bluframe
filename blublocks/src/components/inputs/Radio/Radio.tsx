/* eslint-disable no-ternary */

import { Circle, Input, Label, Wrapper } from "./styled"
import { ComponentProps } from "."

const Radio = ({
  inputId,
  isChecked,
  label,
  labelId,
  name,
  onClick,
  value
}: ComponentProps) => (
  <Wrapper onClick={onClick}>
    <Circle
      aria-checked={isChecked}
      aria-labelledby={labelId}
      checked={isChecked}
      id={inputId}
      role="radio"
    />

    <Input defaultChecked={isChecked} name={name} value={value} />

    <Label id={labelId}>{label}</Label>
  </Wrapper>
)

export default Radio
