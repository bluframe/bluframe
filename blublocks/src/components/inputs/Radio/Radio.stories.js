// @flow

import Radio, { type Props } from "components/inputs/Radio"
import React from "react"
import type { Story } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const stories = {
  component: Radio,
  title: "Inputs/Radio"
}

const Template = (args: Props) => <Radio {...args} />

const args = {
  inputId: "apple-radio",
  isChecked: true,
  label: "Apple",
  name: "fruit",
  onChange: action("onChange"),
  value: "apple"
}

export const Checked: Story<Props> = Template.bind({})
Checked.args = args

export const Unchecked: Story<Props> = Template.bind({})
Unchecked.args = {
  ...args,
  isChecked: false
}

export default stories
