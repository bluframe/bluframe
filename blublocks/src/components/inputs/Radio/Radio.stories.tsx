import Radio from "components/inputs/Radio"
import { StoryObj } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const meta = {
  component: Radio,
  title: "Inputs/Radio"
}

type Story = StoryObj<typeof Radio>

const args = {
  inputId: "apple-radio",
  isChecked: true,
  label: "Apple",
  name: "fruit",
  onChange: action("onChange"),
  value: "apple"
}

export const Checked: Story = {
  args
}

export const Unchecked: Story = {
  args: {
    ...args,
    isChecked: false
  }
}

export default meta
