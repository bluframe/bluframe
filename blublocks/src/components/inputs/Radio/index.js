// @flow

import Radio from "./Radio"
import { prepareComponent } from "@bluframe/grapple"

export type Props = {|
  +inputId?: string,
  +isChecked: boolean,
  +label: string,
  +name: string
|}

export type ComponentProps = {|
  ...Props,
  +labelId?: string
|}

const usePrepareComponent = (props: Props): ComponentProps => {
  const labelId = props.inputId && `${props.inputId}-label`

  return {
    ...props,
    labelId
  }
}

export default (prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Radio
): React$ComponentType<Props>)
