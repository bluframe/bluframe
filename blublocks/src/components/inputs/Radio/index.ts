import React, { MouseEventHandler } from "react"
import Radio from "./Radio"
import { prepareComponent } from "@bluframe/grapple"

export type OptionValue = number | string

export interface Props {
  inputId?: string
  isChecked: boolean
  label: React.ReactNode
  name: string
  // eslint-disable-next-line no-unused-vars
  onChange: (value: OptionValue) => void
  value: OptionValue
}

export interface ComponentProps extends Omit<Props, "onChange"> {
  labelId?: string
  onClick: MouseEventHandler<HTMLElement>
}

const usePrepareComponent = (props: Props): ComponentProps => {
  const labelId = props.inputId && `${props.inputId}-label`

  const onClick = () => {
    props.onChange(props.value)
  }

  return {
    ...props,
    labelId,
    onClick
  }
}

export default prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  Radio
)
