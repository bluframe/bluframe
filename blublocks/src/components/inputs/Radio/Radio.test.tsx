// @flow

import type { ComponentProps } from "."
import Radio from "./Radio"
import React from "react"
import { render } from "tests"

jest.mock(".", () => ({}))

describe("Radio", () => {
  const props: ComponentProps = {
    inputId: "apple-radio",
    isChecked: true,
    label: "Apple",
    name: "Fruit",
    onClick: jest.fn(),
    value: "apple"
  }

  it("renders", () => {
    const { container } = render(<Radio {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
