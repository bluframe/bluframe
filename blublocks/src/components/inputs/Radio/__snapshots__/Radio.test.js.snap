// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders 1`] = `
<div
  class="styled__Wrapper-sc-xszf5l-0 ChdSW"
>
  <div
    aria-checked="true"
    class="styled__Circle-sc-xszf5l-1 dlVTcX"
    id="apple-radio"
    role="radio"
  />
  <input
    checked=""
    class="styled__Input-sc-xszf5l-2 fdhJkT"
    name="Oat Milk"
    type="radio"
  />
  <label
    class="styled__Label-sc-xszf5l-3 fpblxE"
    id="apple-radio-label"
  >
    Apple
  </label>
</div>
`;
